src/coreclr/tools/Common/TypeSystem/IL/Stubs/PInvokeILEmitter.cs

```diff
@@ -97,7 +97,9 @@ private static Marshaller[] InitializeMarshallers(MethodDesc targetMethod, Inter
                 TypeDesc parameterType = (i == 0)
```

Comment:
`(i == 0 && isHRSwapping)` condition is repeated multiple times here. Also, I do not think you should need the check for `DelegateMarshallingMethodThunk` here. We should make sure to set `PreserveSig=true` for marshalled delegates earlier (`UnmanagedFunctionPointerAttribute` won't allow you to specify `PreserveSig=false` so we do not need to worry about handling it).

I think it would be a bit more efficient and readable like this: 
```
TypeDesc parameterType;
bool isHRSwappedRetVal = false;
if (i == 0)
{
    //first item is the return type
    parameterType = methodSig.ReturnType;
    if (!flags.PreserveSig && !parameterType.IsVoid)
    {
        // PreserveSig = false can only show up an regular forward PInvokes
        Debug.Assert(direction == MarshalDirection.Forward);
    
        parameterType = methodSig.Context.GetByRefType(parameterType);
        isHRSwappedRetVal = true;
    }
}
else
{
    parameterType = methodSig[i - 1];
}
```